Gain control of macro namespace visibility

khwilliamson opened the pull request:
This p.r. adds the capability to undefine macros that are visible to XS code but shouldn't be.  This can be used to stop macro namespace pollution by perl.

It works by changing embed.h to have two modes, controlled by a #ifdef that is set by perl.h.  perl.h now #includes embed.h twice.  The first time works as it always has.  The second sets the #ifdef, and causes embed.h to #undef the macros that shouldn't be visible.  This call is just before perl.h returns to its includer, so that these macros have come and gone before the file that #included perl.h is affected by them.  It comes after the inline headers get included, so they have access to all the symbols that are defined.

The list of macros is determined by the visibility given by the apidoc lines documenting them, plus several exception lists that allow a symbol to be visible even though it is not documented as such.

In this p.r., the main exception list contains everything that is currently visible outside the Perl core, so this should not break any code.  But it means that  visibility control is established for future changes to our code base.  New macros will not be visible except when documented as needing to be such.  We can no longer inadvertently add new names to pollute the user's.

I expect that over time, the exception list will become smaller, as we go through it and remove the items that really shouldn't be visible.  We can then see via smoking if someone is actually using them, and either decide that these should be visible, or work with the module author for another way to accomplish their needs.  (I would hope this would lead to proper documentation of the ones that need to be visible.)

There are currently four lists of symbols.

One list is for symbols that are used by libc functions, and that Perl may redefine (usually so that code doesn't have to know if it is running on a platform that is lacking the given feature.)  The algorithm added here catches most of these and keeps them visible, but there are a few items that currently must be manually listed.

A second list is of symbols that the re extension to Perl requires, but no one else needs to.  This list is currently empty, as everything initially is in the main exception list.

A third list is for items that other Perl extensions require, but no one else needs to.  This list is currently empty, as everything initially is in the main exception list.

The final list is for items that currently are visible to the whole world.  It contains thousands of items.  This list should be examined for:

1.  Names that shouldn't be so visible; and
2. Names that need to remain visible but should be changed so they are less likely to clash with anything the user might come up  with.

I have wanted this ability to happen for a long time; and now things have come together to enable it.

This allows us to have a clear-cut boundary with CPAN.

It means you can add macros that have internal-only use without having to worry about making them likely not to clash with user names.

It shows precisely in one place what our names are that are visible to CPAN.


* This set of changes requires a perldelta entry, and I need help writing it.

khwilliamson committed:
regen/embed.pl: Don't create prototype for private macros

Prior to this commit, a commented-out prototype was created for all
macros that we have argument and return type information for.  This
might be useful information for a reader of proto.h.

This commits stops doing this for private macros.  It makes a future
commit slightly easier, and I'm unsure of the usefulness of this anyway.
But I could be persuaded otherwise.

khwilliamson committed:
regen/embed.pl: Don't create prototype for weird macros

Where 'weird' is defined as meaning something where the normal rules
don't apply, so something we generate is unlikely to be correct.

This only affects one element, and it uses aTHX in a way that is
incompatible with it being automated.

khwilliamson committed:
mathoms.c: Add a few coments

khwilliamson committed:
embed.pl: enum and union are like struct in parameters

These keywords all need another word to indicate the parameter type.
Previously only 'struct' was considered to have.

This changed showed an error in one entry embed.fnc, which is also
corrected in this commit.

khwilliamson committed:
embed.fnc: Drop Perl_ on do_aexec my_stat my_lstat

These macros are not for external use, so don't need a Perl_ prefix

khwilliamson committed:
Move hv_stores() declaration from embed.fnc to hv.h

This is required for the next few commits that start automatically
creating long Perl_name functions for the elements in embed.fnc that are
macros and don't already have them in the source.

Only macros can take a parameter that has to be a literal string, so
don't fit with the next few commits.  This is the only case in embed.fnc
like that, so I'm deferring dealing with it for now.

khwilliamson committed:
Remove const char * const return value

A function can't return something of that type, but this has always been
a macro, so this hasn't been caught.

khwilliamson committed:
Three macros are used in perl extensions

They were marked as core-only.

khwilliamson committed:
regen/embed.pl: Move declaration

Future commits make the new location a better choice

khwilliamson committed:
regen/embed.pl: Move declaration to first use

khwilliamson committed:
Gain control of macro namespace visibility

This commit adds the capability to undefine macros that are visible to
XS code but shouldn't be.  This can be used to stop macro namespace
pollution by perl.

It works by changing embed.h to have two modes, controlled by a #ifdef
that is set by perl.h.  perl.h now #includes embed.h twice.  The first
time works as it always has.  The second sets the #ifdef, and causes
embed.h to #undef the macros that shouldn't be visible.  This call is
just before perl.h returns to its includer, so that these macros have
come and gone before the file that #included perl.h is affected by them.
It comes after the inline headers get included, so they have access to
all the symbols that are defined.

The list of macros is determined by the visibility given by the apidoc
lines documenting them, plus several exception lists that allow a symbol
to be visible even though it is not documented as such.

In this commit, the main exception list contains everything that is
currently visible outside the Perl core, so this should not break any
code.  But it means that the visibility control is established for
future changes to our code base.  New macros will not be visible except
when documented as needing to be such.  We can no longer inadvertently
add new names to pollute the user's.

I expect that over time, the exception list will become smaller, as we
go through it and remove the items that really shouldn't be visible.  We
can then see via smoking if someone is actually using them, and either
decide that these should be visible, or work with the module author for
another way to accomplish their needs.  (I would hope this would lead to
proper documentation of the ones that need to be visible.)

There are currently four lists of symbols.

One list is for symbols that are used by libc functions, and that Perl
may redefine (usually so that code doesn't have to know if it is running
on a platform that is lacking the given feature.)  The algorithm added
here catches most of these and keeps them visible, but there are a few
items that currently must be manually listed.

A second list is of symbols that the re extension to Perl requires, but
no one else needs to.  This list is currently empty, as everything
initially is in the main exception list.

A third list is for items that other Perl extensions require, but no one
else needs to.  This list is currently empty, as everything initially is
in the main exception list.

The final list is for items that currently are visible to the whole
world.  It contains thousands of items.  This list should be examined
for:

    1) Names that shouldn't be so visible; and
    2) Names that need to remain visible but should be changed so they
       are less likely to clash with anything the user might come up
       with.

I have wanted this ability to happen for a long time; and now things
have come together to enable it.

This allows us to have a clear-cut boundary with CPAN.

It means you can add macros that have internal-only use without having
to worry about making them likely not to clash with user names.

It shows precisely in one place what our names are that are visible to
CPAN.

jkeenan commented:
> This p.r. undefines all macros that are visible to XS code but shouldn't be. This stops macro namespace pollution by perl.

Can you cite some examples of where "we" (Perl programmers in general, or, at a minimum, CPAN authors) have gotten into trouble with this?

khwilliamson commented:
Getting control of our boundaries is a big deal.  That's why we have EXPORT in perl code.

In the past, we've run into trouble with people using interfaces that weren't supposed to be public, and then complaining if we changed them.  The first big step to controlling that that I can think of is @xenu adding the ability to restrict visibility of non-public functions.  I fairly recently extended that to functions that are visible to perl extensions, though I think there are bypasses to that if someone really wants to.  This step extends this to macros.   Again, a determined programmer can defeat these.  But it's harder to do.

But we have willy-nilly created symbols that infringe on the XS writer's name space.  Writers have had to work around this at times.  There are modules that people can load that try to undo this.  This p.r. has us undoing much of that intrusion.
See  https://stackoverflow.com/questions/22903542/what-is-namespace-pollution.

A recent example of a minor issue is https://github.com/Perl/perl5/pull/23614#issuecomment-3466025887
This p.r. causes that to be a non-issue, for the moment anyway.

We as a project need to have a discussion about what sorts of symbols can intrude.  I have always operated under the assumption that we could reserve symbols ending in an underscore for our use.  I don't know where I got that idea; probably because others who preceded me had made that assumption.  But this needs to be clarified.,

This p.r. brings out into the open for examination just what macros are visible to XS writers.

tonycoz commented:
> A recent example of a minor issue is [#23614 (comment)](https://github.com/Perl/perl5/pull/23614#issuecomment-3466025887)
> This p.r. causes that to be a non-issue, for the moment anyway.

It doesn't.

Setting `PERL_DO_UNDEFS` makes the perl API essentially unusable.

As a quick example it undefs SvSetSV_and(), which is used by the API SvSetSV() and SvSetMagicSV() macros.

It also undefs many flag check macros and accessors that are API (even if not in embed.fnc) and are accessible as functions, eg `CVf_METHOD`, most CV flag checks, GvIOn(), 

My complaint in the above comment was that the `o1_` and `type1_` macros are too short and too global, which could be fixed by making OpTYPE_set() an inline function as @leont suggested, or as I described in that comment.  This change nukes from orbit.

khwilliamson commented:
I have revised this p.r. to not undefine anything that wasn't already visible to the outside world, and changed the description.  It by itself shouldn't break anything, but set us up for gaining control of the namespace.

tonycoz commented on util.h:
```diff
@@ -252,7 +252,7 @@ returning NULL if not found.  The terminating NUL bytes are not compared.
 
 
-#define instr(haystack, needle) strstr(haystack, needle)
+#define instr(haystack, needle) strstr((char *) haystack, (char *) needle)
```
Exactly what message and where in the source inspired this change?

Also, instr() is probably obsolete - left over from pre-ANSI C builds of perl (strstr() is C89)

khwilliamson replied:
I can't reproduce this now; so will remove the commit that changed it.

tonycoz commented on embed.h:
```diff
@@ -21,77 +21,23 @@
+#if defined(PERL_DO_UNDEFS)
+# if !defined(PERL_CORE)
+#   undef do_aexec
+#   undef free_c_backtrace
```
Nothing defines `free_c_backtrace`, only `Perl_free_c_backtrace` is defined.

khwilliamson replied:
This is a bug in embed.fnc.  I'll add a commit to fix it

tonycoz commented on perl.h:
```diff
@@ -9354,6 +9354,16 @@ END_EXTERN_C
 */
 
+START_EXTERN_C
+
```
Does this need `extern "C"`?  embed.h only defines macros and the `#include "embed.h"` ~~at the top of the file~~ around line 6249 doesn't have one.

https://github.com/Perl/perl5/blob/f7dbd626639ae46f762ffa53230fd0c51cc8cab7/perl.h#L6246-L6251

khwilliamson replied:
I cargo culted it from somewhere.  I'll try taking it out

tonycoz commented on regen/embed.pl:
```diff
@@ -59,7 +69,5440 @@ BEGIN
+# For all modules that aren't deliberating using particular names, all the
+# other symbols on it are namespace pollutants.
+
+my @unexpectedly_used_outside_core = qw(
```
`@unexpectedly_used_outside_core` seems like a bad name, many symbols are API and expected to be used outside of core (like `AMGf_assign`) and many don't seem to be used outside of core at all.

khwilliamson replied:
I want to convey that being on this list is not good.  Being here is a sign that something is defective elsewhere.  For example, if `AMGf_assign` were properly documented, it wouldn't need to be here, because the regen software would pick it up.  (It turns out that in this case, it is documented, just not marked as such to `autodoc`.  So it's an easy fix to mark it; I'll issue a separate p.r. for that.)

I'm changing the name to `overrides_to_be_visible_outside_core` but am certainly open to other suggestions.

tonycoz replied:
Some names listed aren't visible outside of code, like `ref` (I suspect it was last visible over 20 years ago).  I haven't looked for others.

Many of them shouldn't be visible outside of core (like the `/^case_\d+_SBOX32$/` names).

And there's a bunch which should be visible outside of core.

Maybe `@unresolved_visibility_overrides`, since ideally the list will become empty as:
- stuff is documented
- we undef stuff like the `case_128_SBOX32` names
- the header scanning code catches up

khwilliamson replied:
Changed to your proposed name

I am working on catching the header scanning code up.  So far it removes about 1500 symbols from the list that aren't actually visible outside core.  It doesn't yet catch `ref`

khwilliamson committed:
embed.fnc: free_c_backtrace is only Perl_free_c_backtrace

khwilliamson commented:
I wanted to make sure this p.r. didn't actually change anything so it can't possibly break anything.  But the code is there.  To check, I just now removed `ref` from the override list, and regenerated.  I got the following diff in embed.h
```
 26c26,30
 < #if !defined(PERL_DO_UNDEFS)
 ---
 > #if defined(PERL_DO_UNDEFS)
 > # if !defined(PERL_CORE)
 > #   undef ref
 > # endif
 > #else
```

khwilliamson committed:
Macros guarded by some #ifdef's aren't globally visible

The previous commit undefines macros that aren't supposed to be visible
to XS code.  But, to avoid any possible breakage, it creates an
exception list of symbols that may have been visible, and leaves them
so.  The goal is to stop the list from growing as new code is developed,
and to shorten the list by various means.

This is the first commit to do that, by looking to see if any symbols
aren't actually externally visible because they are guarded by #ifdef's
that evaluate to false.  For example a symbol that is #defined only if
PERL_CORE is defined won't be visible, and need not be on the exception
list.

This cuts almost 30% off the initial list.

khwilliamson committed:
Remove a few more macros from being visible to XS code

These are a few macros dealing with inversion lists that were never
intended to be visible to general XS code, and they actually can't be in
use in cpan because the mechanisms to create inversion lists are private
to perl.

khwilliamson committed:
Remove SBOX case statements from external visibility

I'm pretty sure there is no use case for these, and very unlikely to
have any actual uses.

tonycoz commented on embed.h:
```diff
@@ -21,77 +21,275 @@
+ * earlier when this file was #included with this symbol undefined */
+#if defined(PERL_DO_UNDEFS)
+# if !defined(PERL_CORE)
+#   undef case_100_SBOX32
+#   undef case_101_SBOX32
+#   undef case_102_SBOX32
+#   undef case_103_SBOX32
+#   undef case_104_SBOX32
+#   undef case_105_SBOX32
+#   undef case_106_SBOX32
+#   undef case_107_SBOX32
+#   undef case_108_SBOX32
+#   undef case_109_SBOX32
```
I think these SBOX32 undefs belong at the tail of the header that defines them, but I'm don't think the code in HeaderParser and embed.pl will recognise the names have been undefined.

khwilliamson replied:
HeaderParser does lay out undefs.  But embed.pl doesn't look at those.  I'm leery of getting into situations where the same symbol is defined and undefined and redefined, and maybe gets defined under some set of conditions and undefined under an overlapping set.  It can get complicated and error prone.

And it may work here to manually undefine these SBOX ones  in the header; or their scope may need to be larger.  We'd have to try it.  But we have thousands of symbols.  This is automatic, and scope isn't an issue.

tonycoz commented:
It does seem to clean up the symbols, and I don't think it should break APIs as is.

I'm not entirely comfortable with the bulk undefs, not every API gets well tested or comprehensively documented, and I can see this undefining symbols down the road that breaks those APIs in the wild.

But I don't see an alternative either.

tonycoz reviewed (approved):
As you said in IRC, this shouldn't be hard to revert if we need to, and the earlier it's in the earlier we find problems.
